opm/simulators/wells/ParallelWellInfo.cpp

```diff
@@ -0,0 +1,325 @@
+                                   [[maybe_unused]] Communication allComm)
+    : name_(well_info.first), hasLocalCells_(well_info.second),
+      rankWithFirstPerf_(-1)
+{
+#if HAVE_MPI
+    MPI_Comm newComm;
+    int color = hasLocalCells_ ? 1 : MPI_UNDEFINED;
+    MPI_Comm_split(allComm, color, allComm.rank(), &newComm);
+    comm_.reset(new Communication(newComm));
+#else
+    comm_.reset(new Communication(Dune::MPIHelper::getLocalCommunicator()));
+#endif
+    commAbove_.reset(new CommunicateAbove(*comm_));
+    isOwner_ = (comm_->rank() == 0);
+}
+
+void ParallelWellInfo::communicateFirstPerforation(bool hasFirst)
+{
+    int first = hasFirst;
+    std::vector<int> firstVec(comm_->size());
+    comm_->allgather(&first, 1, firstVec.data());
+    auto found = std::find_if(firstVec.begin(), firstVec.end(),
+                              [](int i) -> bool{ return i;});
+    rankWithFirstPerf_ = found - firstVec.begin();
+}
+
+void ParallelWellInfo::pushBackEclIndex(int above, int current)
+{
+    commAbove_->pushBackEclIndex(above, current);
+}
+
+void ParallelWellInfo::beginReset()
+{
+    commAbove_->beginReset();
+}
+
+
+void ParallelWellInfo::endReset()
+{
+    commAbove_->beginReset();
+}
+
+void ParallelWellInfo::clearCommunicateAbove()
+{
+    commAbove_->clear();
+}
+
+std::vector<double> ParallelWellInfo::communicateAboveValues(double zero_value,
+                                                             const double* current_values,
+                                                             std::size_t size) const
+{
+    return commAbove_->communicate(zero_value, current_values,
+                                   size);
+}
+
+std::vector<double> ParallelWellInfo::communicateAboveValues(double zero_value,
+                                                             const std::vector<double>& current_values) const
+{
+    return commAbove_->communicate(zero_value, current_values.data(),
+                                   current_values.size());
+}
+
+bool operator<(const ParallelWellInfo& well1, const ParallelWellInfo& well2)
+{
+    return well1.name() < well2.name() || (! (well2.name() < well1.name()) && well1.hasLocalCells() < well2.hasLocalCells());
+}
+
+bool operator==(const ParallelWellInfo& well1, const ParallelWellInfo& well2)
+{
+    bool ret = well1.name() == well2.name() && well1.hasLocalCells() == well2.hasLocalCells()
+        && well1.isOwner() == well2.isOwner();
+#if HAVE_MPI
+    using MPIComm = typename Dune::MPIHelper::MPICommunicator;
+    ret = ret &&
+        static_cast<MPIComm>(well1.communication()) == static_cast<MPIComm>(well2.communication());
+#endif
+    return ret;
+}
+
+bool operator!=(const ParallelWellInfo& well1, const ParallelWellInfo& well2)
+{
+    return ! (well1 == well2);
+}
+
+bool operator<(const std::pair<std::string, bool>& pair, const ParallelWellInfo& well)
+{
+    return pair.first < well.name() || ( !( well.name() < pair.first ) && pair.second < well.hasLocalCells() );
```

Comment:
Interesting idea. To make this work, `hasLocalCells()` would need to return `const bool&`, and then the line would read:
```
return static_cast<const std::tuple<std::string, bool>>(pair) < std::tie(well.name(), well.hasLocalCells());
```
I am not sure whether this is easier to read for everybody.

Reply:
I would only change if somebody really insists on it. To me this change would not improve readability.

Reply:
And back to start it is...

We need to find wells by name that perforate local cells (ìt is `hasLocalCells()`). The load balancer returns pairs, and we need to compare.

> I still think it is better to store the info in a map by name instead of a sorted vector, which adds complexity.

I don't. this saves storage and should have better performance. Usage is rather confined, too.